Update 0x203-C3-Oracle.md

CDSecurity opened the pull request:
Added 7 considerations when using Chainlink Oracle, as well as the link to the article with further explaination.

damianrusinek commented on 2.0/0x100-General/0x106-G6-Communications.md:
```diff
@@ -21,7 +21,8 @@ Category “G6” lists requirements related to the function calls between the v
 | **G6.7** | Verify that the result of low-level function calls (e.g. *send*, *delegatecall*, *call*) from another contracts is checked. | 
-| **G6.8** | Verify that contract relies on the data provided by the right sender and the contract does not rely on *tx.origin* value. | 
+| **G6.8** | Verify that contract relies on the data provided by right sender and contract does not rely on tx.origin value. | 
+| **G6.9** | Verify that contract does not enforce usage of "phantom functions" |
```
Please follow formatting (period at the end).

damianrusinek commented on 2.0/0x100-General/0x108-G8-Denial-of-Service.md:
```diff
@@ -23,6 +23,7 @@ Category “G8” lists requirements related to the possible denial of service o
 | **G8.10** | Verify that if there is a possibility of suspending the operation of the contract, it is also possible to resume it. | 
 | **G8.11** | Verify that if allow lists and deny lists are used, it does not interfere with the normal operation of the system. | 
 | **G8.12** | Verify that there is no DoS caused by overflows and underflows. | 
+| **G8.13** | Don't rely on the contract 0 balance, as it can be forced to receive Ether via coinbase or selfdestruct. |
```
This is already in Business Logic category. Why do you suggest to copy it here?

damianrusinek commented on 2.0/0x100-General/0x109-G9-Blockchain-Data.md:
```diff
@@ -20,6 +20,8 @@ Category “G9” lists requirements related to the blockchain data of the smart
 | **G9.3** | Verify that no confidential data is stored in the blockchain (passwords, personal data, token etc.). | 
 | **G9.4** | Verify that contract does not use string literals as keys for mappings. Verify that global constants are used instead to prevent Homoglyph attack. | 
 | **G9.5** | Verify that contract does not generate pseudorandom numbers trivially basing on the information from blockchain (e.g. seeding with the block number). | 
+| **G9.6** | Don't rely on block timestamp for precise operations, as they can be manipulated by miner. |
```
I am not aware of any real risk associated to 15 seconds spread of block timestamp. Could you elaborate?

deliriusz replied:
For example, using block.timestamp as a source of randomness (it happens to this day)

damianrusinek commented on 2.0/0x200-Components/0x201-C1-Token.md:
```diff
@@ -23,6 +23,7 @@ Category “C1” lists requirements related to the Token smart contract as one
 | **C1.7** | Verify that the transfer business logic is consistent, especially when re-sending tokens to the same address (*from == to*). |
 | **C1.8** | Verify that if the implemented functions include external calls, they are handled correctly and do not introduce unsafe external calls to the system. |
 | **C1.9** | Verify that the *approve()* function from the ERC-20 standard change the allowed amount only to *0* or from *0*. |
+| **C1.10** | Make sure that the token does not have double entry point and does not track user balances in more than one smart contract. |
```
I would suggest to change it to form similar to: Verify that the protocol handles double-entry tokens (tracking user balances in a contract represented by two addresses) correctly or forbids them.

damianrusinek commented on 2.0/0x200-Components/0x203-C3-Oracle.md:
```diff
@@ -20,6 +20,8 @@ Category “C3” lists requirements related to the Oracle smart contract as one
 | **C3.3** | Verify that there are alerts set and monitored for large and sudden changes in the price feed. |
 | **C3.4** | Verify if there is a way to mark the data as incorrect. |
 | **C3.5** | Verify that the supply with incorrect data is penalized. |
+| **C3.6** | Verify that the oracle implements Time Weighted Average Price (TWAP) funcitonality. |
```
TWAP is one type of oracles (price oracles) and this category considers building an oracle, not integrating with it.
I would suggest a check in Oracle Integration category that would propose to consider using TWAP for price on-chain oracles.

damianrusinek commented on 2.0/0x200-Components/0x203-C3-Oracle.md:
```diff
@@ -20,6 +20,8 @@ Category “C3” lists requirements related to the Oracle smart contract as one
 | **C3.4** | Verify if there is a way to mark the data as incorrect. |
 | **C3.5** | Verify that the supply with incorrect data is penalized. |
+| **C3.6** | Verify that the oracle implements Time Weighted Average Price (TWAP) funcitonality. |
+| **C3.7** | Verify that the oracle price for an asset cannot be influenced accross single block. |
```
Verify that the price for an asset returned by oracle cannot be influenced in a single block.

damianrusinek commented on 2.0/0x200-Components/0x204-C4-Vault.md:
```diff
@@ -23,10 +23,17 @@ Category “C4” lists requirements related to the Vault smart contract as one
 | **C4.7** | Verify that the same amount of shares for different users allows the same amount of funds to be withdrawn. |
 | **C4.8** | Verify that a user depositing the same amount earlier will get more rewards than the one who deposits the same amount later. |
+| **C4.9** | Verify that depositing rounds down the assets and withdrawing rounds up. |
+| **C4.10** | If the vault is supposed to support fee-on-transfer tokens, make sure to check for balance change rather than amount to be transferred. |
```
Verify that the transferred amount is confirmed by checking balances before and after the deposit if the vault is going to support fee-on-transfer tokens (or any token is allowed).

damianrusinek commented on 2.0/0x200-Components/0x204-C4-Vault.md:
```diff
@@ -23,10 +23,17 @@ Category “C4” lists requirements related to the Vault smart contract as one
 | **C4.8** | Verify that a user depositing the same amount earlier will get more rewards than the one who deposits the same amount later. |
+| **C4.9** | Verify that depositing rounds down the assets and withdrawing rounds up. |
+| **C4.10** | If the vault is supposed to support fee-on-transfer tokens, make sure to check for balance change rather than amount to be transferred. |
+| **C4.11** | Verify that the vault is only allowed to transfer tokens from msg.sender, to prohibit stealing from other users who set the allowance for the contract |
```
Verify that the vault is allowed to transfer tokens only from _msg.sender_ to prohibit stealing from users who approved the vault contract.

damianrusinek commented on 2.0/0x200-Components/0x204-C4-Vault.md:
```diff
@@ -23,10 +23,17 @@ Category “C4” lists requirements related to the Vault smart contract as one
+| **C4.9** | Verify that depositing rounds down the assets and withdrawing rounds up. |
+| **C4.10** | If the vault is supposed to support fee-on-transfer tokens, make sure to check for balance change rather than amount to be transferred. |
+| **C4.11** | Verify that the vault is only allowed to transfer tokens from msg.sender, to prohibit stealing from other users who set the allowance for the contract |
+| **C4.11** | Verify that the deposit and withdraw business logic is consistent and simmetrical, especially when re-sending tokens to the same address (from == to). |
```
symmetrical

damianrusinek commented on 2.0/0x200-Components/0x205-C5-Bridge.md:
```diff
@@ -23,7 +23,11 @@ Category “C5” lists requirements related to the Bridge smart contract as one
 | **C5.7** | Verify that signed message cannot be used in a different context (use domain separator from EIP-712). |
 | **C5.8** | Verify that the case of 0 being returned by ecrecover function is handled securely. |
-
+| **C5.9** | Verify that priviledged contracts are separated from cross-chain relay calls. |
```
Could you elaborate what you mean here?

deliriusz replied:
Sure, this was the exact attack on Poly Network - https://research.kudelskisecurity.com/2021/08/12/the-poly-network-hack-explained/ . The attacker managed to escalate privileges via a function name hash collision on a relayed message. The attack was possible because relaying contract contained privileged function registering keepers, which could funnel the user funds out the of protocol.

damianrusinek commented on 2.0/0x300-Integrations/0x303-I3-Oracle.md:
```diff
@@ -17,8 +17,10 @@ Category “I3” lists requirements related to the Oracle smart contract as one
+| **I3.3** | Verify that Oracle data is up-to-date. |
+| **I3.4** | Verify that no spot oracle is used (e.g. spot price from Uniswap pool). |
+| **I4.4** | Verify that, when using Uniswap V3 TWAP as price oracle, liquidity is high enough and is distributed widely accross most of the price range. |
+| **I4.5** | Whenever possible, use decentralized offchain oracles unsucceptible to price manipulatino attacks (e.g. Chainlink), ideally adding a fallback onchain oracle. |
```
typos
I am not sure that off-chain oracles is a good solution "whenever possible". They also have multiple downsides. And those are not always decentralized :)

deliriusz replied:
Ok, I changed to "Consider using a decentralized offchain oracles unsucceptible to price manipulation attacks (e.g. Chainlink) for low liquidity asset, ideally adding a fallback onchain oracle." - what do you think?

damianrusinek replied:
We have addressed it and proposed a final check :)

deliriusz replied:
Done in https://github.com/ComposableSecurity/SCSVS/pull/10/commits/6dc31b45da4c9b07243c21f7d7dadb8b1a92bd61

damianrusinek reviewed (commented):
Thanks for contributing! 
I've left some comments.

damianrusinek commented on 2.0/0x300-Integrations/0x303-I3-Oracle.md:
```diff
@@ -17,8 +17,10 @@ Category “I3” lists requirements related to the Oracle smart contract as one
+| **I3.3** | Verify that Oracle data is up-to-date. |
+| **I3.4** | Verify that no spot oracle is used (e.g. spot price from Uniswap pool). |
+| **I4.4** | Verify that, when using Uniswap V3 TWAP as price oracle, liquidity is high enough and is distributed widely across most of the price range. |
+| **I4.5** | Consider using a decentralized off-chain oracles unsusceptible to price manipulation attacks (e.g. Chainlink) for low liquidity asset, ideally adding a fallback on-chain oracle. |
```
```suggestion
| **I4.5** | Verify that, the use a decentralized off-chain oracles unsusceptible to on-chain price manipulation attacks (e.g. Chainlink) is considered for low liquidity asset, ideally combining it with on-chain oracles to detect malicious values. |
```

damianrusinek reviewed (changes requested):
We have addressed all your questions and suggestions and added some small fixes and suggestions. 
Hopefully, it's the last review round :)

CDSecurity committed:
Update 0x203-C3-Oracle.md

Added 7 considerations when using Chainlink Oracle, as well as the link to the article with further explaination.
